File: fetch-pr-descriptions.py
#!/usr/bin/env python3

import re
import os
import requests
from typing import List, Tuple
import time
from dotenv import load_dotenv

# Load environment variables from .env file
load_dotenv()

def read_changelog(file_path: str) -> str:
    """Read the changelog file content."""
    with open(file_path, 'r') as f:
        return f.read()

def find_pr_references(content: str) -> List[Tuple[str, str]]:
    """Find all PR references in the content and return list of (line, pr_id) tuples."""
    pr_pattern = r'/pull/(\d+)'
    lines = content.split('\n')
    pr_references = []
    
    for line in lines:
        match = re.search(pr_pattern, line)
        if match:
            pr_id = match.group(1)
            pr_references.append((line, pr_id))
    
    return pr_references

def fetch_pr_description(pr_id: str) -> str:
    """Fetch PR description from GitHub API."""
    # Get GitHub token from environment variable
    token = os.getenv('GITHUB_TOKEN')
    if not token:
        raise ValueError("GITHUB_TOKEN environment variable is not set")
    
    headers = {
        'Authorization': f'token {token}',
        'Accept': 'application/vnd.github.v3+json'
    }
    
    url = f'https://api.github.com/repos/woocommerce/woocommerce/pulls/{pr_id}'
    response = requests.get(url, headers=headers)
    
    if response.status_code == 200:
        return response.json()['body']
    elif response.status_code == 403:
        print(f"Rate limit exceeded. Waiting 60 seconds...")
        time.sleep(60)
        return fetch_pr_description(pr_id)
    else:
        print(f"Error fetching PR {pr_id}: {response.status_code}")
        return ""

def extract_changes_section(description: str) -> str:
    """Extract the 'Changes proposed' section from PR description."""
    pattern = r'Changes proposed in this Pull Request:.*?(?=\n\n|\Z)'
    match = re.search(pattern, description, re.DOTALL)
    if match:
        return match.group(0).strip()
    return ""

def update_changelog(content: str, pr_references: List[Tuple[str, str]]) -> str:
    """Update the changelog content with PR descriptions."""
    lines = content.split('\n')
    updated_lines = []
    
    for line in lines:
        updated_lines.append(line)
        for ref_line, pr_id in pr_references:
            if line == ref_line:
                description = fetch_pr_description(pr_id)
                changes = extract_changes_section(description)
                if changes:
                    updated_lines.append(changes)
                    updated_lines.append('')  # Add blank line for readability
    
    return '\n'.join(updated_lines)

def main():
    # Get the changelog file path from command line argument
    if len(sys.argv) != 2:
        print("Usage: python fetch-pr-descriptions.py <changelog_file>")
        sys.exit(1)
    
    changelog_file = sys.argv[1]
    
    # Read the changelog
    content = read_changelog(changelog_file)
    
    # Find PR references
    pr_references = find_pr_references(content)
    
    # Update the changelog with PR descriptions
    updated_content = update_changelog(content, pr_references)
    
    # Write the updated content back to the file
    with open(changelog_file, 'w') as f:
        f.write(updated_content)
    
    print(f"Updated {len(pr_references)} PR references in {changelog_file}")

if __name__ == '__main__':
    import sys
    main()

File: fetch-pr-descriptions.php
#!/usr/bin/env php
<?php
/**
 * Augment changelog files with PR descriptions from GitHub.
 *
 * Usage: php fetch-pr-descriptions.php <changelog_file>
 *
 * This script reads a changelog file, finds PR references, fetches their
 * descriptions from GitHub, and appends the "Changes proposed" section
 * to the changelog.
 */

require_once __DIR__ . '/includes/functions.php';

// Load environment variables
$env = parse_env_file(__DIR__ . '/.env');
$GITHUB_TOKEN = $env['GITHUB_TOKEN'] ?? '';
$REPO_OWNER = 'woocommerce';
$REPO_NAME = 'woocommerce';

/**
 * Read the changelog file content.
 *
 * @param string $file_path The path to the changelog file
 * @return string The file content
 */
function read_changelog($file_path) {
    if (!file_exists($file_path)) {
        throw new Exception("File not found: {$file_path}");
    }
    return file_get_contents($file_path);
}

/**
 * Find all PR references in the content.
 *
 * @param string $content The changelog content
 * @return array Array of [line, pr_id] tuples
 */
function find_pr_references($content) {
    $pr_pattern = '/\/pull\/(\d+)/';
    $lines = explode("\n", $content);
    $pr_references = [];

    foreach ($lines as $line) {
        if (preg_match($pr_pattern, $line, $matches)) {
            $pr_id = $matches[1];
            $pr_references[] = [$line, $pr_id];
        }
    }

    return $pr_references;
}

/**
 * Fetch PR description from GitHub API.
 *
 * @param string $pr_id The PR ID
 * @return string The full PR body
 */
function fetch_pr_description_body($pr_id) {
    global $GITHUB_TOKEN, $REPO_OWNER, $REPO_NAME;

    if (empty($GITHUB_TOKEN)) {
        throw new Exception("GITHUB_TOKEN environment variable is not set");
    }

    $url = "https://api.github.com/repos/{$REPO_OWNER}/{$REPO_NAME}/pulls/{$pr_id}";
    $response = github_request($url, $GITHUB_TOKEN, true);

    if ($response['http_code'] === 200) {
        $pr_data = json_decode($response['body'], true);
        return $pr_data['body'] ?? '';
    } elseif ($response['http_code'] === 403) {
        echo "Rate limit exceeded. Waiting 60 seconds...\n";
        sleep(60);
        return fetch_pr_description_body($pr_id);
    } else {
        echo "Error fetching PR {$pr_id}: HTTP {$response['http_code']}\n";
        return '';
    }
}

/**
 * Extract the "Changes proposed" section from PR description.
 *
 * @param string $description The full PR description
 * @return string The extracted changes section
 */
function extract_pr_changes_section($description) {
    $pattern = '/Changes proposed in this Pull Request:.*?(?=\n\n|\Z)/s';
    if (preg_match($pattern, $description, $matches)) {
        return trim($matches[0]);
    }
    return '';
}

/**
 * Update the changelog content with PR descriptions.
 *
 * @param string $content The original changelog content
 * @param array $pr_references Array of [line, pr_id] tuples
 * @return string The updated changelog content
 */
function update_changelog($content, $pr_references) {
    $lines = explode("\n", $content);
    $updated_lines = [];

    foreach ($lines as $line) {
        $updated_lines[] = $line;

        foreach ($pr_references as list($ref_line, $pr_id)) {
            if ($line === $ref_line) {
                $description = fetch_pr_description_body($pr_id);
                $changes = extract_pr_changes_section($description);
                if (!empty($changes)) {
                    $updated_lines[] = $changes;
                    $updated_lines[] = ''; // Add blank line for readability
                }
            }
        }
    }

    return implode("\n", $updated_lines);
}

// Main execution
if (php_sapi_name() === 'cli') {
    // Get the changelog file path from command line argument
    if ($argc !== 2) {
        echo "Usage: php fetch-pr-descriptions.php <changelog_file>\n";
        exit(1);
    }

    $changelog_file = $argv[1];

    try {
        // Read the changelog
        $content = read_changelog($changelog_file);

        // Find PR references
        $pr_references = find_pr_references($content);

        if (empty($pr_references)) {
            echo "No PR references found in {$changelog_file}\n";
            exit(0);
        }

        // Update the changelog with PR descriptions
        $updated_content = update_changelog($content, $pr_references);

        // Write the updated content back to the file
        file_put_contents($changelog_file, $updated_content);

        echo "Updated " . count($pr_references) . " PR references in {$changelog_file}\n";

    } catch (Exception $e) {
        echo "Error: " . $e->getMessage() . "\n";
        exit(1);
    }
}
